### PR TITLE
Make unexported fields panic (informatively) instead of silently passing

### LINKED ISSUE #3928: Semantic DeepEqual passes objects with unexported fields, when it probably shouldn't
https://github.com/thockin/kubernetes/tree/semantic_eq

This tree has a commit that demonstrates two obviously different objects that should not pass the semantic DeepEqual but do.

@lavalamp

### PR DESCRIPTION
**Pull request by lavalamp**
Closes #3928.

### CODE REVIEW

**Comment by thockin** [comment]
LGTM

**Comment by lavalamp** [comment]
Looks like this fails in places other than the serialization test. Will fix.

**Comment by lavalamp** [comment]
OK well that was a sequence of disturbing findings. Fixed now, I think.

File: pkg/registry/controller/rest_test.go [hunk 1]
```diff
@@ -378,7 +379,7 @@ func TestFillCurrentState(t *testing.T) {
 	if controller.Status.Replicas != 2 {
 		t.Errorf("expected 2, got: %d", controller.Status.Replicas)
 	}
-	if !api.Semantic.DeepEqual(fakeLister.s, labels.Set(controller.Spec.Selector).AsSelector()) {
+	if !reflect.DeepEqual(fakeLister.s, labels.Set(controller.Spec.Selector).AsSelector()) {
```
Comment (thockin) [inline pkg/registry/controller/rest_test.go]: FYI - if we get some concept of "internal contstructors" then maybe we can get rid of this and have versioned APIs just use "string" or "IntOrString" - dumb structs.

**Comment by thockin** [comment]
LGTM

File: pkg/api/helpers.go [hunk 2]
```diff
@@ -60,6 +60,9 @@ var Semantic = conversion.EqualitiesOrDie(
 		}
 		return a.Amount.Cmp(b.Amount) == 0
 	},
+	func(a, b util.Time) bool {
+		return a.Unix() == b.Unix()
```
Comment (thockin) [inline pkg/api/helpers.go]: If we fuzz properly we should be able to use .UTC() here right?

Comment (lavalamp) [inline pkg/api/helpers.go]: Yeah, thanks for changing it.

**Pull request merged**